### PR TITLE
[FIX] {purchase_,}mrp: Various MO Overview fixes

### PR DESCRIPTION
**Pull request by clesgow**
Various fixes of the MO Overview :

1. Adds byproducts section as in the BoM Overview.
2. Add MO MTO links that are already done in the replenishments.
3. Fix the display of quantities "Free to use / On hand" so it correctly checks the warehouse of the Manufacturing Order.
4. Removes extra "In Transit" line when using MTO components.
5. Changes the meaning of cost columns.
    - `MO Cost` means the expected costs following *this* MO.
    - `Product Cost` renamed to `Real Cost` and have it represent what were the actual registered costs for this MO, from a stock point of view (i.e. by using valuation layers if possible). 
6. Removes checks to replenishment rules when the MO is done.
7. Hide unused components when the MO is done.
8. If a replenishment PO has a different currency than the MO's company, correctly converts its value in the report.
9. Fix availability for MTO subcontracted subassemblies.
10. Include the cost_analysis based on already computed data from the report.

Enterprise part : https://github.com/odoo/enterprise/pull/40930

Task-3251506

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by Whenrow** [comment]
robodoo r+

**Comment by Whenrow** [comment]
robodoo r-

File: addons/mrp/report/mrp_report_mo_overview.py [hunk 1]
```diff
@@ -696,7 +696,7 @@ def _get_reserved_qty(self, move_raw, warehouse, replenish_data):
         if not replenish_data['qty_reserved'].get(move_raw):
             total_reserved = 0
             wh_location_ids = self._get_warehouse_locations(warehouse, replenish_data)
-            linked_moves = self.env['stock.move'].browse(move_raw._rollup_move_origs()).filtered(lambda m: m.location_id.id in wh_location_ids or m.location_dest_id.id not in wh_location_ids)
+            linked_moves = self.env['stock.move'].browse(move_raw._rollup_move_origs()).filtered(lambda m: m.location_id.id in wh_location_ids)
```
Comment (yhu-odoo) [inline addons/mrp/report/mrp_report_mo_overview.py:699]: can't we use `not m.is_subcontract`?

File: addons/mrp/report/mrp_report_mo_overview.py [hunk 2]
```diff
@@ -73,10 +77,10 @@ def _get_report_extra_lines(self, summary):
             'unit_product_cost': unit_product_cost,
         }
 
-    def _get_mo_summary(self, production, components, operations):
+    def _get_mo_summary(self, production, components, current_mo_cost, byproducts_cost_portion):
         product = production.product_id
         company = production.company_id or self.env.company
-        mo_cost = sum(component.get('summary', {}).get('mo_cost', 0.0) for component in components)
+        byproducts_cost_share = float_round(1 - byproducts_cost_portion, precision_rounding=0.0001)
```
[the remark this answers is not in the document]
Comment (clesgow) [reply addons/mrp/report/mrp_report_mo_overview.py]: This line is changed in the following commit in the same PR : 81d5f67

File: addons/mrp/report/mrp_report_mo_overview.py [hunk 3]
```diff
@@ -45,13 +47,17 @@ def _get_report_values(self, docids, data=None):
...
+        return production.with_context(warehouse=production.warehouse_id.id)
+
     def _get_display_context(self):
         return {
             'show_uom': self.env.user.user_has_groups('uom.group_uom'),
         }
 
-    def _get_report_data(self, production_id):
-        production = self.env['mrp.production'].browse(production_id)
+    def _get_report_data(self, production):
```
[the remark this answers is not in the document]
Comment (clesgow) [reply addons/mrp/report/mrp_report_mo_overview.py:76]: It's to simplify the override done enterprise-side, where `_get_report_data()` also uses the modified `production` record. This was done avoid searching for it all over again.

File: addons/mrp/report/mrp_report_mo_overview.py (hunk 1 above)
Comment (clesgow) [reply addons/mrp/report/mrp_report_mo_overview.py:699]: One issue I see with this is that `is_subcontract` is defined in the `mrp_subcontracting`, requiring a new override in this module. While I agree this modification is about the subcontracting, it seems a bit overkill to me.
Also, the problematic move is the one from Production -> Subcontracting Location, which is not `is_subcontract` (that would be its dest move)

**Review by ticodoo — COMMENTED** [review]
The `[FIX] mrp: hide components that weren't consumed in done MO Overview` commit needs to be updated because we no longer split the line for the not done qtys, but we still don't want to include non-consumed lines anyways so the fix is still valid (the joys of extra long PRs)

the rest is mostly nitpicks

File: addons/mrp/report/mrp_report_mo_overview.py [hunk 4]
```diff
@@ -486,18 +499,90 @@ def _get_replenishments_from_forecast(self, production, replenish_data):
...
+
+                    move_out_qty -= used_quantity
+                    if float_compare(move_out_qty, 0, line['move_out'].product_uom.rounding) <= 0:
+                        break
+        return new_lines + forecast_lines
+
+    def _get_origin(self, move):
+        if move.production_id:
+            return (move.production_id, move)
+        return (False, False)
```
[the remark this answers is not in the document]
Comment (clesgow) [reply addons/mrp/report/mrp_report_mo_overview.py:766]: You're correct, there is no benefit of doing so. I guess (and hope) that at some point it was useful, but that's clearly not the case anymore :thinking: 
Updated to your suggestion.

File: addons/mrp/report/mrp_report_mo_overview.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (clesgow) [reply addons/mrp/report/mrp_report_mo_overview.py:79]: Update: My previous comment is no longer relevant as most of the enterprise code has been moved.
Moved the context addition to `_get_report_data()`.

File: addons/mrp/report/mrp_report_mo_overview.py [hunk 5]
```diff
@@ -140,17 +235,32 @@ def _get_comparison_decorator(self, expected, current, rounding):
...
+            is_workorder_started = not float_is_zero(wo_duration, precision_digits=2)
+            bom_cost = expected_cost
+            if workorder.operation_id:
+                operation = workorder.operation_id
+                capacity = operation.workcenter_id._get_capacity(production.product_id)
+                operation_cycle = float_round(production.product_uom_qty / capacity, precision_rounding=1, rounding_method='UP')
+                bom_duration_expected = (operation_cycle * operation.time_cycle * 100.0 / operation.workcenter_id.time_efficiency) + \
+                    operation.workcenter_id._get_expected_duration(production.product_id)
+                bom_cost = expected_cost / workorder.duration_expected * bom_duration_expected
+            mo_cost = expected_cost if production.state != 'done' else current_cost
```
[the remark this answers is not in the document]
Comment (clesgow) [reply addons/mrp/report/mrp_report_mo_overview.py:261]: Good catch, thanks!

File: addons/mrp/report/mrp_report_mo_overview.py [hunk 6]
```diff
@@ -304,19 +524,28 @@ def _get_replenishment_lines(self, production, move_raw, replenish_data, level,
...
                 'state': doc_in.state,
                 'formatted_state': self._format_state(doc_in),
-                'quantity': min(quantity, forecast_uom_id._compute_quantity(forecast_line['quantity'], move_raw.product_uom)),  # Avoid over-rounding
+                'quantity': line_quantity,
                 'uom_name': move_raw.product_uom.display_name,
                 'uom_precision': self._get_uom_precision(forecast_line['uom_id']['rounding']),
-                'mo_cost': forecast_line.get('cost', self._get_replenishment_cost(product, forecast_line['quantity'], forecast_uom_id, currency, forecast_line.get('move_in'))),
-                'product_cost': currency.round(forecast_uom_id._compute_quantity(forecast_line['quantity'], product.uom_id) * product.standard_price),
+                'unit_cost': self._get_unit_cost(move_raw),
+                'mo_cost': forecast_line.get('cost', self._get_replenishment_cost(product, line_quantity, forecast_uom_id, currency, forecast_line.get('move_in'))),
```
[the remark this answers is not in the document]
Comment (clesgow) [reply addons/mrp/report/mrp_report_mo_overview.py:531]: Correct, good catch.

File: addons/mrp/report/mrp_report_mo_overview.py [hunk 7]
```diff
@@ -470,7 +470,8 @@ def _get_replenishment_lines(self, production, move_raw, replenish_data, level,
                 'currency': currency,
             }}
             if resupply_data:
-                to_order_line['summary']['mo_cost'] = currency.round(resupply_data['cost'])
+                mo_cost = resupply_data['currency']._convert(resupply_data['cost'], currency, (production.company_id or self.env.company), fields.Date.today())
```
[the remark this answers is not in the document]
Comment (clesgow) [reply addons/mrp/report/mrp_report_mo_overview.py:586]: Agreed, hopefully we can use the cache when it's ready.

File: addons/mrp/report/mrp_report_mo_overview.py [hunk 8]
```diff
@@ -300,6 +342,12 @@ def _format_component_move(self, production, move_raw, replenishments, company,
 
         return component
 
+    def _get_component_real_cost(self, move_raw, quantity):
```
[the remark this answers is not in the document]
Comment (ticodoo) [reply addons/mrp/report/mrp_report_mo_overview.py:458]: As discussed in person, if you ever feel like cleaning up the code, there are 2 lines in `_get_replenishment_lines` that do the same work as this method (except for the check if the qty is 0). This is not important though, so ignoring for this PR

**Comment by ticodoo** [comment]
@robodoo rebase-ff r+

**Pull request closed without merging**